### PR TITLE
Add input validation docs and tooling

### PR DESCRIPTION
**Pull request by tgorgol**
**Description**

Changes proposed in this pull request:

- Add input validation document
- Add tooling for input validation

**Related issue(s)**

#255

### CODE REVIEW

File: components/director/pkg/inputvalidation/directive.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+func (d *directive) Validate(ctx context.Context, obj interface{}, next graphql.Resolver) (res interface{}, err error) {
+	constructedObj, err := next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	validatableObj, ok := constructedObj.(Validatable)
+	if !ok {
+		return nil, errors.New("misuse of directive, object is not validatable")
```
Comment (dbadura) [inline components/director/pkg/inputvalidation/directive.go]:
```suggestion
		return nil, errors.Errorf("misuse of directive, object is not validatable: %T", constructedObj)

```

File: docs/architecture/input-validation.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+--- | --- | ---
+key: String! | `required`, `printable`, `max=256` | varchar(256) in db  
+value: Any! (interface{}) | `required` | 
+
+### FetchRequestInput
+
+Field | Rules | Comment
+--- | --- | ---
+url: String! | `required`, `url`, `printable`, `max=256` | varchar(256) in db  
+auth: AuthInput |  |
```
Comment (dbadura) [inline docs/architecture/input-validation.md]: ? `required`?

**Comment by dbadura** [comment]
We could mark optional parameters with `optional` rule or something similar.
Look at CloudFoundry API doc:
http://docs.cloudfoundry.org/api/uaa/version/74.4.0/#overview
We can do something like this:

| Field | Required | Rules |
|---|---|---|
| name | true | `Name`

File: components/director/pkg/inputvalidation/directive.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	Validate() error
+}
+
+type directive struct{}
+
+func NewDirective() *directive {
+	return &directive{}
+}
+
+func (d *directive) Validate(ctx context.Context, obj interface{}, next graphql.Resolver) (res interface{}, err error) {
```
Comment (crabtree) [inline components/director/pkg/inputvalidation/directive.go]:
```suggestion
func (d *directive) Validate(ctx context.Context, obj interface{}, next graphql.Resolver) (nterface{}, error) {
```

File: components/director/pkg/inputvalidation/directive_test.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+func (ts testStruct) Validate() error {
+	return ts.returnErr
+}
+
+func TestDirective_Validate(t *testing.T) {
+	validatorDirective := NewDirective()
+	testErr := errors.New("testError")
+	ctx := context.TODO()
+	t.Run("Success", func(t *testing.T) {
```
Comment (crabtree) [inline components/director/pkg/inputvalidation/directive_test.go]: to be consistent with other tests

```suggestion
	t.Run("success", func(t *testing.T) {
```

**Comment by dbadura** [comment]
👍

File: docs/architecture/input-validation.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+- `[$VALIDATION_RULE]` - Array that can be nil or empty but every array element has to fulfill specified `$VALIDATION_RULE`.
+
+
+## Proposed validation rules for Compass input types
+
+### APIDefinitionInput
+
+Field | Required | Rules | Comment
+--- | --- | --- | ---
+name: String! | true | `name` |
```
Comment (crabtree) [inline docs/architecture/input-validation.md]: The last `|` is not necessary here

**Pull request merged**